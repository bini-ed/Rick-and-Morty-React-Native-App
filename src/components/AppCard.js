import {
  View,
  StyleSheet,
  Dimensions,
  Image,
  TouchableOpacity,
} from 'react-native';
import React, {memo} from 'react';
import AppText from './AppText';

const {width} = Dimensions.get('screen');

const AppCard = ({item, navigation}) => {
  return (
    <TouchableOpacity
      onPress={() => navigation.navigate('CharacterDetail', {id: item?.id})}
      style={{marginVertical: 10}}>
      <Image
        source={{uri: item?.image}}
        style={{
          width: width / 2 - 20,
          height: 220,
          borderRadius: 15,
          resizeMode: 'contain',
        }}></Image>

      <View style={styles.card}>
        <AppText
          font={{
            fontSize: 15,
            fontFamily: 'serif',
            fontWeight: '600',
            alignSelf: 'center',
            color: '#156900',
          }}>
          {item?.name}
        </AppText>
        <AppText font={{fontSize: 13, alignSelf: 'center'}}>
          {item?.species}, {item?.gender}, {item?.status}
        </AppText>
        <AppText font={{fontSize: 13, alignSelf: 'center'}}>
          {item?.origin?.name}
        </AppText>
      </View>
    </TouchableOpacity>
  );
};

export default memo(AppCard);
const styles = StyleSheet.create({
  card: {
    // marginVertical: 10,
    // backgroundColor: '#0F1A09',
    borderRadius: 10,
    width: width / 2 - 20,
  },
});
