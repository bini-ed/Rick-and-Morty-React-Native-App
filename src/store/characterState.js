import {createSlice} from '@reduxjs/toolkit';
import {Toast} from 'react-native-toast-message/lib/src/Toast';
import {apiCallBegan} from './apiAction';

const slice = createSlice({
  name: 'character',
  initialState: {
    list: [],
    next: null,
    loading: false,
    loadMore: false,
    searchload: false,
  },
  reducers: {
    characterLoading: (character, action) => {
      character.loading = true;
    },
    characterClear: (character, action) => {
      character.list = [];
    },
    characterFetch: (character, action) => {
      character.list = action.payload?.results;
      character.next = action.payload?.info.next;
      character.loading = false;
    },
    characterFetchFailed: (character, action) => {
      character.loading = false;
      Toast.show({type: 'error', text1: action.payload.error});
    },
    characterLoadMore: (character, action) => {
      character.loadMore = true;
    },
    characterFetchMore: (character, action) => {
      action.payload.results.map(items => {
        character.list.push(items);
      });
      character.next = action.payload?.info.next;
      character.loadMore = false;
    },
    characterFetchMoreFailed: (character, action) => {
      character.loadMore = false;
      Toast.show({type: 'error', text1: action.payload.error});
    },
    singleCharacterLoading: (character, action) => {
      character.searchload = true;
    },
    singleCharacterFetch: (character, action) => {
      character.list.push(action.payload);
      character.searchload = false;
    },
    singleCharacterFailed: (character, action) => {
      character.searchload = false;
      Toast.show({type: 'error', text1: action.payload.error});
    },
  },
});

export const {
  characterFetch,
  characterLoading,
  characterFetchFailed,
  characterFetchMore,
  characterFetchMoreFailed,
  characterLoadMore,
  singleCharacterFetch,
  singleCharacterFailed,
  singleCharacterLoading,
  characterClear,
} = slice.actions;
export default slice.reducer;

export const loadCharacters = () =>
  apiCallBegan({
    url: 'character',
    onStart: characterLoading.type,
    onSuccess: characterFetch.type,
    onError: characterFetchFailed.type,
  });

export const loadMoreCharacters = url =>
  apiCallBegan({
    url: url,
    onStart: characterLoadMore.type,
    onSuccess: characterFetchMore.type,
    onError: characterFetchMoreFailed.type,
  });

export const searchCharacters = params =>
  apiCallBegan({
    url: `character/?${params}`,
    onStart: characterLoading.type,
    onSuccess: characterFetch.type,
    onError: characterFetchFailed.type,
  });
export const getSingleCharacters = INCOME_URL =>
  apiCallBegan({
    url: INCOME_URL,
    onStart: singleCharacterLoading.type,
    onSuccess: singleCharacterFetch.type,
    onError: singleCharacterFailed.type,
  });
