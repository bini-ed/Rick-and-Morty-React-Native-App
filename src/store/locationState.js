import {createSlice} from '@reduxjs/toolkit';
import {Toast} from 'react-native-toast-message/lib/src/Toast';
import {apiCallBegan} from './apiAction';

const slice = createSlice({
  name: 'location',
  initialState: {
    list: [],
    next: null,
    loading: false,
    loadMore: false,
  },
  reducers: {
    locationFetchLoading: location => {
      location.loading = true;
    },
    locationFetch: (location, action) => {
      location.list = action.payload?.results;
      location.next = action.payload?.info?.next;
      location.loading = false;
    },
    locationFetchFailed: (location, action) => {
      location.loading = false;
      Toast.show({type: 'error', text1: action.payload.error});
    },
    locationFetchMoreLoading: location => {
      location.loadMore = true;
    },
    locationFetchMore: (location, action) => {
      action.payload.results.map(items => {
        location.list.push(items);
      });
      location.next = action.payload?.info?.next;
      location.loadMore = false;
    },
    locationFetchMoreFailed: (location, action) => {
      location.loading = false;
      Toast.show({type: 'error', text1: action.payload.error});
    },
  },
});

export default slice.reducer;

export const {
  locationFetch,
  locationFetchFailed,
  locationFetchLoading,
  locationFetchMore,
  locationFetchMoreLoading,
  locationFetchMoreFailed,
} = slice.actions;

export const fetchLocation = () =>
  apiCallBegan({
    url: 'location',
    onStart: locationFetchLoading.type,
    onSuccess: locationFetch.type,
    onError: locationFetchFailed.type,
  });

export const fetchMoreLocation = url =>
  apiCallBegan({
    url: url,
    onStart: locationFetchMoreLoading.type,
    onSuccess: locationFetchMore.type,
    onError: locationFetchMoreFailed.type,
  });
