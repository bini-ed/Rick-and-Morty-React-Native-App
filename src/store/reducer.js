import {combineReducers} from '@reduxjs/toolkit';
import characterReducer from './characterState';
import locationReducer from './locationState';

export default combineReducers({
  character: characterReducer,
  location: locationReducer,
});
