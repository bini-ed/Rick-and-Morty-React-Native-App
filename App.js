import {GestureHandlerRootView} from 'react-native-gesture-handler';
import {
  SafeAreaView,
  StatusBar,
  StyleSheet,
  Text,
  useColorScheme,
  View,
} from 'react-native';
import React from 'react';
import {Provider} from 'react-redux';
import Toast from 'react-native-toast-message';
import {NavigationContainer} from '@react-navigation/native';
import store from './src/store/store';
import BottomTabNaviagtion from './src/navigation/BottomTabNaviagtion';
import {Colors} from 'react-native/Libraries/NewAppScreen';

const App = () => {
  const isDarkMode = useColorScheme() === 'dark';

  const backgroundStyle = {
    backgroundColor: isDarkMode ? Colors.darker : Colors.lighter,
    flex: 1,
  };
  return (
    <SafeAreaView style={backgroundStyle}>
      <Provider store={store}>
        <GestureHandlerRootView style={{flex: 1}}>
          <NavigationContainer>
            <StatusBar
              barStyle={isDarkMode ? 'light-content' : 'dark-content'}
              backgroundColor={backgroundStyle.backgroundColor}
            />

            <BottomTabNaviagtion />
          </NavigationContainer>
          <Toast />
        </GestureHandlerRootView>
      </Provider>
    </SafeAreaView>
  );
};

export default App;
