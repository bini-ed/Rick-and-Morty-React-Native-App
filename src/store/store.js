import {configureStore, getDefaultMiddleware} from '@reduxjs/toolkit';
import reducer from '../store/reducer';
import apiCall from './middleware/apiCallMiddleware';

const store = configureStore({
  reducer,
  middleware: [
    ...getDefaultMiddleware({
      immutableCheck: false,
      serializableCheck: false,
    }),
    apiCall,
  ],
});
export default store;
