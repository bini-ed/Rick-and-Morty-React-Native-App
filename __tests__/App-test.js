/* eslint-disable no-console */

import 'react-native';
import React from 'react';
import App from '../App';

// Note: test renderer must be required after react-native.
import AppCard from '../src/components/AppCard';
import renderer, {create} from 'react-test-renderer';
import * as actions from '../src/store/characterState';
import store from '../src/store/store';
import {Provider} from 'react-redux';

describe('Character and Location returns actions', () => {
  const tree = create(
    <Provider store={store}>
      <AppCard />
    </Provider>,
  );
  it('Check App card componenet', async () => {
    expect(tree).toMatchSnapshot();
  });

  it('Test fetch character action dispatch ', async () => {
    expect(store.getState().character).toEqual({
      list: [],
      next: null,
      loading: false,
      loadMore: false,
      searchload: false,
    });
  });
  it('Test fetch location action dispatch ', async () => {
    expect(store.getState().location).toEqual({
      list: [],
      next: null,
      loading: false,
      loadMore: false,
    });
  });
});
