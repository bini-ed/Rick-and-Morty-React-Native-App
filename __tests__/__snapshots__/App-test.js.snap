// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Character and Location returns actions Check App card componenet 1`] = `
<View
  style={
    Object {
      "marginVertical": 10,
    }
  }
>
  <Image
    source={
      Object {
        "uri": undefined,
      }
    }
    style={
      Object {
        "borderRadius": 15,
        "height": 220,
        "resizeMode": "contain",
        "width": 355,
      }
    }
  />
  <View
    style={
      Object {
        "borderRadius": 10,
        "width": 355,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "black",
            "fontFamily": "Avenir",
            "fontSize": 14,
          },
          Object {
            "alignSelf": "center",
            "color": "#156900",
            "fontFamily": "serif",
            "fontSize": 15,
            "fontWeight": "600",
          },
        ]
      }
    />
    <Text
      style={
        Array [
          Object {
            "color": "black",
            "fontFamily": "Avenir",
            "fontSize": 14,
          },
          Object {
            "alignSelf": "center",
            "fontSize": 13,
          },
        ]
      }
    >
      , 
      , 
    </Text>
    <Text
      style={
        Array [
          Object {
            "color": "black",
            "fontFamily": "Avenir",
            "fontSize": 14,
          },
          Object {
            "alignSelf": "center",
            "fontSize": 13,
          },
        ]
      }
    />
  </View>
</View>
`;

exports[`Character and Location returns actions Snapshot 1`] = `
<View
  style={
    Object {
      "marginVertical": 10,
    }
  }
>
  <Image
    source={
      Object {
        "uri": undefined,
      }
    }
    style={
      Object {
        "borderRadius": 15,
        "height": 220,
        "resizeMode": "contain",
        "width": 355,
      }
    }
  />
  <View
    style={
      Object {
        "borderRadius": 10,
        "width": 355,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "black",
            "fontFamily": "Avenir",
            "fontSize": 14,
          },
          Object {
            "alignSelf": "center",
            "color": "#156900",
            "fontFamily": "serif",
            "fontSize": 15,
            "fontWeight": "600",
          },
        ]
      }
    />
    <Text
      style={
        Array [
          Object {
            "color": "black",
            "fontFamily": "Avenir",
            "fontSize": 14,
          },
          Object {
            "alignSelf": "center",
            "fontSize": 13,
          },
        ]
      }
    >
      , 
      , 
    </Text>
    <Text
      style={
        Array [
          Object {
            "color": "black",
            "fontFamily": "Avenir",
            "fontSize": 14,
          },
          Object {
            "alignSelf": "center",
            "fontSize": 13,
          },
        ]
      }
    />
  </View>
</View>
`;

exports[`Check App card componenet 1`] = `
<View
  style={
    Object {
      "marginVertical": 10,
    }
  }
>
  <Image
    source={
      Object {
        "uri": undefined,
      }
    }
    style={
      Object {
        "borderRadius": 15,
        "height": 220,
        "resizeMode": "contain",
        "width": 355,
      }
    }
  />
  <View
    style={
      Object {
        "borderRadius": 10,
        "width": 355,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "black",
            "fontFamily": "Avenir",
            "fontSize": 14,
          },
          Object {
            "alignSelf": "center",
            "color": "#156900",
            "fontFamily": "serif",
            "fontSize": 15,
            "fontWeight": "600",
          },
        ]
      }
    />
    <Text
      style={
        Array [
          Object {
            "color": "black",
            "fontFamily": "Avenir",
            "fontSize": 14,
          },
          Object {
            "alignSelf": "center",
            "fontSize": 13,
          },
        ]
      }
    >
      , 
      , 
    </Text>
    <Text
      style={
        Array [
          Object {
            "color": "black",
            "fontFamily": "Avenir",
            "fontSize": 14,
          },
          Object {
            "alignSelf": "center",
            "fontSize": 13,
          },
        ]
      }
    />
  </View>
</View>
`;

exports[`fetch character actions returns actions Snapshot 1`] = `
<View
  style={
    Object {
      "marginVertical": 10,
    }
  }
>
  <Image
    source={
      Object {
        "uri": undefined,
      }
    }
    style={
      Object {
        "borderRadius": 15,
        "height": 220,
        "resizeMode": "contain",
        "width": 355,
      }
    }
  />
  <View
    style={
      Object {
        "borderRadius": 10,
        "width": 355,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "black",
            "fontFamily": "Avenir",
            "fontSize": 14,
          },
          Object {
            "alignSelf": "center",
            "color": "#156900",
            "fontFamily": "serif",
            "fontSize": 15,
            "fontWeight": "600",
          },
        ]
      }
    />
    <Text
      style={
        Array [
          Object {
            "color": "black",
            "fontFamily": "Avenir",
            "fontSize": 14,
          },
          Object {
            "alignSelf": "center",
            "fontSize": 13,
          },
        ]
      }
    >
      , 
      , 
    </Text>
    <Text
      style={
        Array [
          Object {
            "color": "black",
            "fontFamily": "Avenir",
            "fontSize": 14,
          },
          Object {
            "alignSelf": "center",
            "fontSize": 13,
          },
        ]
      }
    />
  </View>
</View>
`;

exports[`fetch character actions test fetch characters for character page 1`] = `
Object {
  "payload": undefined,
  "type": "character/characterFetch",
}
`;

exports[`fetch character actions test search character for 1`] = `
Object {
  "payload": Object {
    "onError": "character/characterFetchFailed",
    "onStart": "character/characterLoading",
    "onSuccess": "character/characterFetch",
    "url": "character/?undefined",
  },
  "type": "api/CallBegan",
}
`;

exports[`fetch character actions test search character for with params 1`] = `
Object {
  "payload": Object {
    "onError": "character/characterFetchFailed",
    "onStart": "character/characterLoading",
    "onSuccess": "character/characterFetch",
    "url": "character/?[object Object]",
  },
  "type": "api/CallBegan",
}
`;

exports[`fetch characters for character page 1`] = `
Object {
  "payload": undefined,
  "type": "character/characterFetch",
}
`;
